decryptor/postgresql/pg_decryptor.go

```diff
@@ -232,84 +232,105 @@ func (proxy *PgProxy) ProxyClientConnection(errCh chan<- error) {
+		// After tha packet has been observed and possibly modified, forward it to the database.
+		if err := packet.sendPacket(); err != nil {
+			logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorNetworkWrite).
+				WithError(err).Errorln("Can't send packet")
+			errCh <- err
+			return
 		}
-		if changed {
-			packet.ReplaceQuery(newQuery.Query())
+		// If this is a termination packet, we're done here. Signal EOF and stop the proxy.
+		if packet.terminatePacket {
+			errCh <- io.EOF
+			return
 		}
+	}
+}
 
-		censorSpan.End()
+func (proxy *PgProxy) handleClientPacket(packet *PacketHandler, logger *log.Entry) (bool, error) {
+	// Let the protocol observer take a look at the packet, keeping note of it.
+	packetType, err := proxy.protocolState.HandleClientPacket(packet)
+	if err != nil {
+		return false, err
+	}
+	switch packetType {
+	case QueryPacket:
+		// If that's some sort of a packet with query data inside it, go on...
 
-		// Forwared the (modified) query to the database.
-		if err := packet.sendPacket(); err != nil {
-			logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorNetworkWrite).WithError(err).Errorln("Can't send packet")
-			errCh <- err
-			return
+	default:
+		// Forward all other uninteresting packets to the database without processing.
+		return false, nil
+	}
+	query := proxy.protocolState.PendingQuery()
+
+	// Log query text -- if and only if we're in debug mode -- without inserted value data.
+	// The query can still be sensitive though, so only in debug mode can we do this.
+	if logging.GetLogLevel() == logging.LogDebug {
+		_, queryWithHiddenValues, _, err := common.HandleRawSQLQuery(query.Query())
+		if err == common.ErrQuerySyntaxError {
+			logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorQueryParseError).
+				Debugf("Parsing error on query: %s", queryWithHiddenValues)
+		} else {
+			logger.WithField("sql", queryWithHiddenValues).Debugln("New query")
 		}
 	}
+
+	// Let AcraCensor take a look at the query text.
+	// If it's not okay (and we're still alive), don't let the database see the query.
+	if censorErr := proxy.censor.HandleQuery(query.Query()); censorErr != nil {
+		logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorQueryIsNotAllowed).
+			WithError(censorErr).Errorln("AcraCensor blocked query")
+		return true, nil
+	}
+
+	// Let the registered observers observe the query, potentially modifying it (e.g., transparent encryption).
+	newQuery, changed, err := proxy.queryObserverManager.OnQuery(query)
+	if err != nil {
+		logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorEncryptQueryData).
+			Errorln("Error occurred on query handler")
+	}
+	if changed {
+		packet.ReplaceQuery(newQuery.Query())
+	}
+	return false, nil
+}
+
+func (proxy *PgProxy) sendClientAcraCensorError(logger *log.Entry) error {
+	errorMessage, err := NewPgError("AcraCensor blocked this query")
```

Comment:
Sure, this error won't happen too often. But consider making it constant to avoid generation of the same packet every time it happens. Though I'm not sure whether Go allows calling such complex functions when creating constants (as opposed to C++ `constexpr` of Rust `const fn`).

Reply:
> Though I'm not sure whether Go allows calling such complex functions when creating constants

Don't worry, it does not allow that :) Usually the errors are defined as global *variables*, precisely for that reason. The error objects have identities, so they can't be expressed in terms of simple constants (which work like substitution in Go), and the compiler won't do smart things like initializing them once and replacing uses with references.

Reply:
Woops. It seems that we both got bamboozled by this "error" word.

For some reason I thought that `NewPgError` is just a fancy constructor for Go's `error` when in fact it creates a PostgreSQL packet with a given error. In this case I don't think it's common to extract those things into reusable constants as they are not really useful anywhere else outside of the package. Plus, `NewPgError` can fail and there's a question of how this failure [during program startup] should be handled. It's not nice to just ignore it because we know it won't fail.

So... I've reverted the change (which I quickly applied without even trying to compile), and I think we can tolerate this 'inefficiency' since this condition should not happen often.